### PR TITLE
parsers: add tree-sitter-mlir parser

### PR DESCRIPTION
**Pull request by artagnon**
The parser is currently in development, and 60-80% of MLIR files in the test suite of MLIR, within the Arith, SCF, Tensor, and Linalg dialects parse successfully.

### CODE REVIEW

**Comment by amaanq** [comment]
Can you add queries as well?

**Comment by amaanq** [comment]
Since you noted its in development, please add "experimental = true" in parsers.lua

File: README.md [hunk 1]
```diff
@@ -173,6 +173,7 @@ We are looking for maintainers to add more parsers and to write query files for
 
 <!--This section of the README is automatically updated by a CI job-->
 <!--parserinfo-->
+
```
Comment (theHamsta) [inline README.md:176]: auto-generation of readme will remove this blank-line anyway

File: queries/mlir/highlights.scm [hunk 2]
```diff
@@ -0,0 +1,262 @@
...
+  "linalg.pooling_nwc_min_unsigned"
+  "linalg.pooling_nwc_sum"
+  "linalg.quantized_batch_matmul"
+  "linalg.quantized_matmul"
+  "linalg.vecmat"
+  "linalg.generic"
+  "linalg.index"
+  "linalg.map"
+  "linalg.yield"
+] @function.builtin
```
Comment (theHamsta) [inline queries/mlir/highlights.scm:191]: I don't know whether the usual distinction of control flow/retunr statements makes sense for LLVM IR

File: queries/mlir/highlights.scm [hunk 3]
```diff
@@ -0,0 +1,262 @@
...
+
+[
+  "="
+  "->"
+] @operator
+
+(func_dialect name: (symbol_ref_id) @function)
+(llvm_dialect name: (symbol_ref_id) @function)
+
+(func_arg_list (value_use) @variable.parameter)
```
Comment (theHamsta) [inline queries/mlir/highlights.scm:257]: we have just `@parameter`

File: queries/mlir/highlights.scm (hunk 2 above)
Comment (artagnon) [reply queries/mlir/highlights.scm:191]: It makes sense for LLVM IR, but it's not so clear for MLIR. As per [LangRef](https://mlir.llvm.org/docs/LangRef/), all of them are just custom operations.

File: queries/mlir/highlights.scm [hunk 4]
```diff
@@ -0,0 +1,332 @@
...
+  (type_alias_def)
+] @type
+
+[
+  (integer_literal)
+  (float_literal)
+] @number
+
+[
+  (bool_literal)
```
Comment (theHamsta) [inline queries/mlir/highlights.scm:277]: `@boolean`?

File: queries/mlir/highlights.scm [hunk 5]
```diff
@@ -0,0 +1,332 @@
...
+  (complex_literal)
+  (tensor_literal)
+  (unit_literal)
+] @constant.builtin
```
Comment (theHamsta) [inline queries/mlir/highlights.scm:281]: Usually `@constant` applies for identifiers, not for literals. Complex numbers might just be `@number`

Comment (artagnon) [reply queries/mlir/highlights.scm:281]: How should I classify `tensor_literal`, `unit_literal`, and the like?

**Review by theHamsta — APPROVED** [review]

**Review by theHamsta — APPROVED** [review]

**Comment by theHamsta** [comment]
Thank you!

**Pull request merged**